rust/src/interpreter.rs

```diff
@@ -385,10 +385,11 @@ impl<'a> Interpreter<'a> {
     /// If the const generic S is false, the step check is skipped.
     /// If the const generic J is false jumpdests are skipped.
     #[cfg(not(feature = "tail-call"))]
-    pub fn run<O: Observer, const STEP_CHECK: bool, const JUMPDEST: bool>(
-        &mut self,
-        observer: &mut O,
-    ) -> Result<(), FailStatus> {
+    pub fn run<O, R, const STEP_CHECK: bool, const JUMPDEST: bool>(mut self, observer: &mut O) -> R
```

Comment:
I am not really sure if making the return type generic helps readability. After all the prototype of a function is the best documentation you can get. This prototype invites to get "creative"

Reply:
We have to support returning ExecutionResult and StepResult so I see no way around the generic. Maybe a comment explaining this would help. I will add one.